hasktorch/src/Torch/Typed/Autograd.hs

```diff
@@ -5,38 +5,60 @@
 {-# LANGUAGE TypeFamilies #-}
 {-# LANGUAGE TypeOperators #-}
 {-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE UndecidableInstances #-}
 
-module Torch.Typed.Autograd where
+module Torch.Typed.Autograd
+  ( Torch.Typed.Autograd.HasGrad
+  , Torch.Typed.Autograd.grad
+  )
+where
 
 import           Data.HList
 import           GHC.TypeLits
 import           System.IO.Unsafe
 
-import qualified ATen.Cast as ATen
-import qualified ATen.Class as ATen
+import qualified ATen.Cast                     as ATen
+import qualified ATen.Class                    as ATen
 import qualified Torch.Managed.Autograd
 import qualified Torch.DType                   as D
 import qualified Torch.Device                  as D
 import qualified Torch.Tensor                  as D
 import           Torch.Typed.Tensor
 import           Torch.Typed.Parameter
 
-type family GradR (parameters :: [a]) (dtype :: D.DType) (device :: (D.DeviceType, Nat)) :: [a] where
-  GradR '[] _ _ = '[]
-  GradR (Parameter device dtype shape ': parameters) dtype device = Tensor device dtype shape ': GradR parameters dtype device
-
--- | calculate gradients of a zero-dimensional tensor with respect to a list of parameters
-grad
-  :: forall dtype device parameters gradients tensors
-   . ( gradients ~ GradR parameters dtype device
-     , tensors ~ gradients
-     , HMap' ToDependent parameters tensors
-     , ATen.Castable (HList gradients) [D.ATenTensor]
-     )
-  => Tensor device dtype '[]
-  -> HList parameters
-  -> HList gradients
-grad loss inputs = unsafePerformIO $ ATen.cast2
-  Torch.Managed.Autograd.grad
-  loss
-  (hmap' ToDependent inputs)
+class HasGrad a b | a -> b where
```

Comment:
This is going to be a pretty foundational typeclass. Do you have any other implementations in mind beyond Parameter and HList?

Reply:
I discussed with @apaszke about, in essence, having an instance for Tensor, too. He thinks having it is an easy way to shoot oneself in the foot, for me the benefits of not having to insist on Parameter outweigh that. libtorch's grad can definitely handle it. if we want it later, it's in this file too, just commented for now. other than that, I don't see any other instances for now.